editoast/src/views/authz.rs

```diff
@@ -83,163 +122,624 @@ async fn whoami(Extension(auth): AuthenticationExt) -> Result<Json<WhoamiRespons
+                    result.insert(
+                        Resource::Infra,
+                        vec![ResourceGrant {
+                            resource_type: Resource::Infra,
+                            resource_id: *infra_id,
+                            grant: InfraGrant::Owner,
+                        }],
+                    );
+                }
+            }
+        }
+    }
+
+    Ok(Json(result))
 }
 
-async fn check_user_exists(user_id: i64, regulator: &Regulator) -> Result<()> {
-    if !regulator
-        .user_exists(user_id)
-        .await
-        .map_err(AuthzError::from)?
-    {
-        return Err(NoSuchUserError::NoSuchUser { user_id }.into());
-    }
+#[derive(Clone, Debug, Serialize, Deserialize, ToSchema, PartialEq)]
+struct SubjectItem {
+    pub id: i64,
+    pub name: String,
+    #[schema(inline)]
+    pub r#type: Subject,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, ToSchema, PartialEq)]
+struct SubjectGrant {
+    #[schema(inline)]
+    subject: SubjectItem,
+    #[schema(inline)]
+    grant: InfraGrant,
+}
+
+#[derive(Deserialize, IntoParams)]
+struct ResourceTypeParam {
+    resource_type: Resource,
+}
 
-    Ok(())
+#[derive(Deserialize, IntoParams)]
+struct ResourceIdParam {
+    resource_id: i64,
 }
 
 #[utoipa::path(
-    get, path = "",
+    get,
+    path = "",
     tag = "authz",
-    params(UserIdPathParam),
+    params(ResourceTypeParam, ResourceIdParam, PaginationQueryParams),
     responses(
-        (status = 200, description = "List the roles of a user", body = inline(Roles)),
+        (status = 200, description = "Get list of user that have access to the resource", body = inline(Vec<SubjectGrant>)),
     ),
 )]
-async fn list_user_roles(
-    Path(UserIdPathParam { user_id }): Path<UserIdPathParam>,
-    Extension(auth): AuthenticationExt,
-    State(AppState { regulator, .. }): State<AppState>,
-) -> Result<Json<Roles>> {
-    if !auth
-        .check_roles([Role::OperationalStudies].into())
-        .await
-        .map_err(AuthorizationError::from)?
-    {
-        return Err(AuthorizationError::Forbidden.into());
+async fn users_grants_for_resource_id(
+    State(AppState {
+        regulator, db_pool, ..
+    }): State<AppState>,
+    Path(ResourceTypeParam { resource_type }): Path<ResourceTypeParam>,
+    Path(ResourceIdParam { resource_id }): Path<ResourceIdParam>,
+    Query(pagination_params): Query<PaginationQueryParams>,
+) -> Result<Json<Vec<SubjectGrant>>> {
+    // Checking if the resource type is valid
+    // For now we only support infra
+    if resource_type != Resource::Infra {
+        return Err(ResourceError::UnknownResourceType {
+            resource_type: resource_type.to_string(),
+        })?;
     }
 
-    check_user_exists(user_id, &regulator).await?;
+    // Check that infra exists
+    let infra_id = resource_id;
+    let conn = &mut db_pool.get().await?;
+    Infra::retrieve_or_fail(conn, infra_id, || InfraApiError::NotFound { infra_id }).await?;
```

Comment:
```suggestion
    Infra::exists_or_fail(conn, infra_id, || InfraApiError::NotFound { infra_id }).await?;
```

Not sure if it's worth checking that anymore since we trust the result of the regulator. We can match on the `Err` of the regulator call to return this specific error.